jobs/gtfs-rt-parser/gtfs_rt_parser.py

```diff
@@ -129,51 +117,75 @@ def identify_files(glob, rt_file_type: RTFileType, progress=False) -> List[RTFil
 
 # Originally this whole function was retried, but tmpdir flakiness will throw
 # exceptions in backoff's context, which ruins things
-def parse_file(bucket: str, rt_file: RTFile, pbar=None):
-    fs = get_fs()
+def parse_and_aggregate_hour(bucket: str, hour: RTAggregation, pbar=None):
+    def log(*args, fg=None, **kwargs):
+        # capture fg so we don't pass it to pbar
+        if pbar:
+            pbar.write(*args, **kwargs)
+        else:
+            typer.secho(*args, fg=fg, **kwargs)
 
-    with tempfile.TemporaryDirectory() as tmp_dir:
-        parsed = parse_pb(rt_file.path, open_func=fs.open)
+    fs = get_fs()
 
-        if not parsed or "entity" not in parsed:
-            msg = f"WARNING: no records found in {str(rt_file.path)}"
-            if pbar:
-                pbar.write(msg)
-                pbar.update(1)
-            else:
-                typer.secho(msg, fg=typer.colors.YELLOW)
-            return
+    out_path = f"{hour.hive_path}{EXTENSION}"  # probably a better way to do this
 
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        fs.get(
+            rpath=[file.path for file in hour.source_files],
+            lpath=[os.path.join(tmp_dir, file.path) for file in hour.source_files],
+        )
         gzip_fname = str(tmp_dir + "/" + "temporary" + EXTENSION)
         written = 0
-
         with gzip.open(gzip_fname, "w") as gzipfile:
-            for record in parsed["entity"]:
-                record.update(
-                    {
-                        "header": parsed["header"],
-                        "metadata": json.loads(
-                            rt_file.json()
-                        ),  # back and forth so we use pydantic serialization
-                    }
-                )
-                gzipfile.write((json.dumps(record) + "\n").encode("utf-8"))
-                written += 1
-
-        out_path = f"{rt_file.hive_path(bucket)}{EXTENSION}"  # probably a better way to do this
-        msg = f"writing {written} lines from {str(rt_file.path)} to {out_path}"
-        if pbar:
-            pbar.write(msg)
+            for rt_file in hour.source_files:
+                feed = gtfs_realtime_pb2.FeedMessage()
+
+                try:
+                    with open(os.path.join(tmp_dir, rt_file.path), "rb") as f:
+                        feed.ParseFromString(f.read())
+                    parsed = json_format.MessageToDict(feed)
+                except DecodeError:
```

Comment:
Do we want to just add an error file in this refactor? Re #1429

Reply:
I know I said that I'm not sure we need all the history, but if you're going to run the backfill anyway....?

Reply:
Mixed feelings but yes, probably.... otherwise we'd have to backfill yet again. I'll come up with something.

Reply:
Could be a good chance to establish a pattern for metadata/error files in the new paradigm. If you think it's too much of a headache, we can defer, but seems like it would be nice to just do it now if we can.

Reply:
I've added a generic "RT processing error" concept in https://github.com/cal-itp/data-infra/pull/1423 that gets uploaded in the same manner as other files.